File: actito_flutter/android/src/main/kotlin/com/actito/flutter/events/ActitoEvent.kt
package com.actito.flutter.events

import com.actito.models.ActitoApplication
import com.actito.models.ActitoDevice

internal sealed class ActitoEvent {

    abstract val type: Type
    abstract val payload: Any?

    enum class Type(val id: String) {
        READY(id = "ready"),
        UNLAUNCHED(id = "unlaunched"),
        DEVICE_REGISTERED(id = "device_registered"),
        URL_OPENED(id = "url_opened"),
    }

    class Ready(
        application: ActitoApplication
    ) : ActitoEvent() {
        override val type = Type.READY
        override val payload = application.toJson()
    }

    class Unlaunched : ActitoEvent() {
        override val type = Type.UNLAUNCHED
        override val payload: Nothing? = null
    }

    class DeviceRegistered(
        device: ActitoDevice
    ) : ActitoEvent() {
        override val type = Type.DEVICE_REGISTERED
        override val payload = device.toJson()
    }

    class UrlOpened(
        url: String,
    ) : ActitoEvent() {
        override val type = Type.URL_OPENED
        override val payload = url
    }
}

File: actito_flutter/ios/Classes/Events/ActitoEvent.swift
//
//  ActitoEvent.swift
//  actito
//
//  Created by Helder Pinhal on 03/12/2020.
//

import Foundation
import ActitoKit

enum ActitoEventType: String, CaseIterable {
    case ready
    case unlaunched
    case deviceRegistered = "device_registered"
    case urlOpened = "url_opened"
}

protocol ActitoEvent {
    var type: ActitoEventType { get }
    var payload: Any? { get }
}

class ActitoEventOnReady: ActitoEvent {
    let type: ActitoEventType
    let payload: Any?

    init(application: ActitoApplication) {
        self.type = .ready
        self.payload = try! application.toJson()
    }
}

class ActitoEventOnUnlaunched: ActitoEvent {
    let type: ActitoEventType
    let payload: Any?

    init() {
        self.type = .unlaunched
        self.payload = nil
    }
}

class ActitoEventOnDeviceRegistered: ActitoEvent {
    let type: ActitoEventType
    let payload: Any?

    init(device: ActitoDevice) {
        self.type = .deviceRegistered
        self.payload = try! device.toJson()
    }
}

class ActitoEventOnUrlOpened: ActitoEvent {
    let type: ActitoEventType
    let payload: Any?

    init(url: String) {
        self.type = .urlOpened
        self.payload = url
    }
}
